File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/fortran/LeandroCFD.f90
program OPERADORES_ESTRUCTURAS
    implicit none
    
    !Se declaran las variables a utilizar en el programa
    integer   :: a,aa,ab 
    real      :: b,ba,bb
    complex   :: c,ca,cb
    logical   :: d,da,db
    character (len=60):: e,ea,eb,ec

    !El operador más basico de Fortran es el operador de asignación de valor, tiene la siguiente sintaxis.
    a=1
    b=2.1
    c=(1,2)
    d=.true.
    e='Esto es una prueba'
    print*,a,b,c,d,e

    !OPERACIONES ARITMÉTICAS
    !Fortran tiene operadores aritmeticos de adicción, sustracción, multiplicación, división, potenciación y cambio de signo.
    !Las operaciones aritmeticas solo son permitidas para variables de tipo integer, real y complex
    print*,''
    print*,'***********************'
    print*,'OPERACIONES ARITMETICAS'
    print*,'***********************'
    !Operaciones aritmeticas entre numeros enteros
    print*,''
    print*,'1)Operaciones numeros enteros'
    aa=2;ab=5
    print*,'Entero aa es igual a:',aa,'Entero ab es igual a:',ab
    a=aa+ab !Operador aritmetico de adicción
    print*,'La suma de aa y ab es igual a:',a
    a=aa-ab !Operador aritmetico de sustracción
    print*,'La resta de aa y ab es igual a:',a
    a=aa*ab !Operador aritmetico de multiplicación
    print*,'La multiplicacion de aa y ab es igual a:',a
    a=ab/aa !Operador aritmetico de división
    print*,'La division de aa y ab es igual a:',a
    a=aa**ab !Operador arimetico de potenciación
    print*,'La potenciacion de aa y ab es igual a:',a
    a=-a    !Operador arimetico de cambio de signo
    print*,'El cambio de signo de a es igual a:',a

    !Operaciones aritmeticas entre numeros reales
    print*,''
    print*,'2)Operaciones numeros reales'
    ba=4.3;bb=3.1
    print*,'Entero ba es igual a:',ba,'Entero bb es igual a:',bb
    b=ba+bb !Operador aritmetico de adicción
    print*,'La suma de ba y bb es igual a:',b
    b=ba-bb !Operador aritmetico de sustracción
    print*,'La resta de ba y bb es igual a:',b
    b=ba*bb !Operador aritmetico de multiplicación
    print*,'La multiplicacion de ba y bb es igual a:',b
    b=ba/bb !Operador aritmetico de división
    print*,'La division de ba y bb es igual a:',b
    b=ba**bb !Operador arimetico de potenciación
    print*,'La potenciacion de ba y bb es igual a:',b
    b=-b    !Operador arimetico de cambio de signo
    print*,'El cambio de signo de b es igual a:',b

    !Operaciones aritmeticas entre numeros complejos
    print*,''
    print*,'3)Operaciones numeros complejos'
    ca=(4.3,1);cb=(3.1,2)
    print*,'Entero ca es igual a:',ca,'Entero cb es igual a:',cb
    c=ca+cb !Operador aritmetico de adicción
    print*,'La suma de ca y cb es igual a:',c
    c=ca-cb !Operador aritmetico de sustracción
    print*,'La resta de ca y cb es igual a:',c
    c=ca*cb !Operador aritmetico de multiplicación
    print*,'La multiplicacion de ca y cb es igual a:',c
    c=ca/cb !Operador aritmetico de división
    print*,'La division de ca y cb es igual a:',c
    c=ca**cb !Operador arimetico de potenciación
    print*,'La potenciacion de ca y cb es igual a:',c
    c=-c    !Operador arimetico de cambio de signo
    print*,'El cambio de signo de c es igual a:',c

    !Existe la posibilidad de convertir tipos de variables en otros.
    b=a !Se convierte un número entero (a) en un número real (b), aunque el caso contrario es permitido NO es aconsejable.
    print*,b
    c=b !Se convierte un número real (b) en un número complejo (c)
    print*,c

    !Las operaciones aritmeticas en Fortran tienen un orden de solución:
    !1) Operaciones entre parentesis
    !2) Operaciones de potenciación
    !3) Operaciones de multiplicación y división, yendo de izquierda a derecha
    !4) Operaciones de cambio de signo
    !5) Operaciones de adicción y sustracción, yendo de izquierda a derecha

    !OPERACIONES DE COMPARACIÓN
    !Las operaciones de comparación en Fortran tienen dos sintaxis, una para las versiones 90 o posteriores y una para la versión 77
    !Las expresiones a comparar deben ser de tipo numerico (integer, real o complex) y el resultado es de tipo logical.
    print*,''
    print*,'**************************'
    print*,'OPERACIONES DE COMPARACION'
    print*,'**************************'
    print*,''
    d=ab==aa !Operación de comparación 'es igual a', sintaxis Fortran 90
    d=ab.eq.aa !Operación de comparación 'es igual a', sintaxis Fortran 77
    print*,'ab es igual a aa:',d
    d=ab/=aa !Operación de comparación 'no es igual a', sintaxis Fortran 90
    d=ab.ne.aa !Operación de comparación 'no es igual a', sintaxis Fortran 77
    print*,'ab NO es igual a aa:',d
    d=ab>aa !Operación de comparación 'es estrictamente mayor a', sintaxis Fortran 90
    d=ab.gt.aa !Operación de comparación 'es estrictamente mayor a', sintaxis Fortran 77
    print*,'ab es estrictamente mayor a aa:',d
    d=ab>=aa !Operación de comparación 'es mayor o igual a', sintaxis Fortran 90
    d=ab.ge.aa !Operación de comparación 'es mayor o igual a', sintaxis Fortran 77
    print*,'ab es mayor o igual a aa:',d
    d=ab<aa !Operación de comparación 'es estrictamente menor a', sintaxis Fortran 90
    d=ab.lt.aa !Operación de comparación 'es estrictamente menor a', sintaxis Fortran 77
    print*,'ab es estrictamente menor a aa:',d
    d=ab<=aa !Operación de comparación 'es menor o igual a', sintaxis Fortran 90
    d=ab.le.aa !Operación de comparación 'es menor o igual a', sintaxis Fortran 77
    print*,'ab es menor o igual a aa:',d

    !En este punto aprendí a representar acentos desde la terminal (Se debe cambiar la pagina de codigos a UTF-8, se puede desde 
    !vscode o desde la terminal ejecutando el comando "chcp 65001")

    !OPERACIONES LÓGICAS
    !Las operaciones lógicas solo admiten expresiones de tipo logical y el resultado también es de tipo logical.
    print*,''
    print*,'*******************'
    print*,'OPERACIONES LÓGICAS'
    print*,'*******************'
    print*,''
    d=.false.
    db=.true.
    print*,'d es igual a:',d,'y db es igual a:',db
    da=.not.d !Operador de negación, cambia el valor de la expresión lógica a su opuesto, necesita de una sola expresión.
    print*,'La negacion de d es igual a:',da
    da=d.and.db !Operador de conjunción (y), verdadero si ambas expresiones son verdaderas 
    print*,'d y db es igual a:',da
    da=d.or.db !Operador de disyunción inclusiva (o), verdadero si una de las expresiones es verdadera
    print*,'d o db es igual a:',da
    da=d.eqv.db !Operador de equivalencia (eqv), verdadero si ambas expresiones tienen el mismo valor 
    print*,'d eqv db es igual a:',da
    da=d.neqv.db !Operador de no equivalencia o de disyunción inclusiva (neqv o xor), verdadero si ambas expresiones no tienen el
    !mismo valor 
    print*,'d neqv db es igual a:',da

    !OPERADORES CON CARACTERES
    !Fortran tiene diferentes operaciones con caracteres como los operadores binarios de concatenación y comparación
    print*,''
    print*,'**************************'
    print*,'OPERACIONES CON CARACTERES'
    print*,'**************************'
    print*,''
    ea='   Bogotá,   '
    eb=' 2600 metros más cerca de las estrellas'
    e=adjustl(ea) !Operador que permite eliminar todos los espacios al inicio de la cadena de caracteres
    print*,'Comparación de el operador adjustl: sin adjustl:',ea,'con adjustl:',e
    e=adjustr(ea) !Operador que permite eliminar todos los espacios al final de la cadena de caracteres
    print*,'Comparación de el operador adjustr: sin adjustr:',ea,' con adjustr:',e
    ec=trim(ea) !Operador que permite eliminar todos los espacios al final de la cadena de caracteres de forma temporal
    print*,'Comparación de el operador trim: con trim:',trim(ea),' sin trim:',ea
    e=trim(ea)//trim(eb) !Operador de concatenación
    print*,e
    e=adjustl(trim(ea)//trim(eb)) !Combinación de operadores
    print*,e
    e=ea(5:8) !Parte de candena de caracteres, en este caso "e" queda con los caracteres desde la posición 5 hasta la 8 de "ea"
    print*,e
    e=ea(:6)  !Parte de candena de caracteres, en este caso "e" queda con los caracteres desde el inicio hasta la posición 6 de ea
    print*,e
    e=ea(6:)  !Parte de candena de caracteres, en este caso "e" queda con los caracteres desde 6 hasta el final de "ea"
    print*,e

    !Fortran no tiene operadores de pertenencia como "in" o de identidad como "is".

    !ESTRUCTURAS DE CONTROL
    !CONDICIONALES
    !Fortran permite los condicionales con la función IF, la sintaxis es la siguiente.
    print*,''
    print*,'*************'
    print*,'CONDICIONALES'
    print*,'*************'
    print*,''
    if (ba>bb) aa=10  !Es posible utilizar condicionales en una sola linea de codigo si las instrucciones son sencillas.
    print*,"Resultado condicional una sola linea de codigo:",aa

    if (ba>bb) then   !O en varias lineas de codigo si son varias instrucciones.
        aa=a*5
        print*,"Resultado condicional varias lineas de codigo:",aa        
    end if

    if (bb>ba) then
        aa=a/2
        print*,"Este resultado se imprimira si bb es mayor que ba:",aa
    else              !La instrucción else permite ejecutar instrucciones si la condición del if no se cumple.
        aa=a*2
        print*,"Este resultado se imprimira si bb es menor que ba:",aa        
    end if

    if (bb>ba) then
        aa=a**2
    else if (bb<ba) then  !La instrucción else if permite adicionar un condicional si el primer if no se cumple
        aa=a+ab
        print*,"Este resultado se imprimira si bb es menor que ba (else if):",aa
    else              
        aa=a*2
        print*,"Este resultado se imprimira si bb es igual que ba:",aa   
    end if

    Prueba : if (ba>bb) then   !Se le puede agregar un nombre al if para dar mayor claridad, en este caso "Prueba"
        aa=a*5
        print*,"Resultado condicional varias lineas de codigo:",aa        
    end if Prueba

    !BUCLES O CICLOS
    !Fortran permite diferentes tipos de bucles con la función DO, permite bucles infinitos, bucles while y bucles con 
    !numero finito de iteraciones.
    print*,''
    print*,'******'
    print*,'BUCLES'
    print*,'******'
    print*,''
    !BUCLE INFINITO
    do        
        print*,"Bucle infinito do"
        exit  !Se debe utilizar la instrucción exit para detener el ciclo, exit pasara a la siguiente instrucción luego del end do
    end do
    aa=9    
    bucle : do !También es posible darle un nombre al bucle como un identificador
        !Se llega aquí con la instrucción cycle bucle
        print*,"Bucle infinito do con nombre 'bucle'"
        bucle_dos : do !Es posible anidar bucles infinitos
            !Se llega aquí con la instrucción cycle
            aa=aa+1
            print*,"Bucle infinito do con nombre 'bucle_dos'"
            if (aa==10) cycle !La instrucción cycle obliga al programa a regresar al inicio de las instrucciones del bucle
            if (aa==11) cycle bucle !Cycle se puede asosciar al nombre de un bucle
            if (aa==12) exit bucle_dos !Exit se puede asosciar al nombre de un bucle
            if (aa==13) exit bucle
        end do bucle_dos
        print*,"Se llega aquí con la instrucción exit bucle_dos"
    end do bucle
    print*,"Se llega aquí con la instrucción exit bucle"

    !BUCLE DO WHILE
    do while (aa>10) !Este ciclo continua hasta que aa sea menor o igual a 10
        print*,"Bucle do while, valor de aa es igual a:",aa
        aa=aa-1
    end do

    !BUCLE CON NÚMERO FIJO DE ITERACIONES
    aa=0
    do a = 1,6,2 !Este bucle va desde 1 hasta 6 con un incremento de 2, por lo que solo seran 3 iteraciones.
        aa=aa+1
        print*,"Iteración número:",aa,"El valor de a es igual a:",a
    end do
    print*,"Segundo bucle con número fijo de iteraciones"
    aa=0
    do a = 1,6 !Si no se le asigna ningún incremento este será de 1.
        aa=aa+1
        print*,"Iteración número:",aa,"El valor de a es igual a:",a
    end do

    !ESTRUCTURA CASE
    !La estructura de control case permite seleccionar un bloque de instrucciones a ser ejecutada, la sintaxis es la siguiente. 
    print*,''
    print*,'***************'
    print*,'ESTRUCTURA CASE'
    print*,'***************'
    print*,''
    
    print*,"Digita un valor del 1 hasta el 10:"
    read*,aa
    select case (aa) !La variable entre parentesis puede ser de tipo integer, logical o character
        case (1) 
            print*,"Seleccionaste el valor 1"
        case (2:5) 
            print*,"Seleccionaste un valor del 2 al 5"
        case (6:9) 
            print*,"Seleccionaste un valor del 6 al 9"
        case (10) 
            print*,"Seleccionaste el valor 10"
    end select

    print*,''
    print*,'****************'
    print*,'DIFICULTAD EXTRA'
    print*,'****************'
    print*,''
    
    do a = 10, 55
        aa=mod(a,2)   !La función mod devuelve el resto de una división con resto, si mod es cero la división es exacta.
        Impresion : if (aa==0 .or. a==55) then
            aa=mod(a,3)
            if (aa==0) exit Impresion
            if (a==16) exit Impresion
            print*,a            
        end if Impresion  
    end do

end program OPERADORES_ESTRUCTURAS

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/c++/LeandroCFD.cpp
#include <iostream>
#include <string>


int main() {

    // Declaración de variables
    int a, b, i;
    float c, d;
    bool e, f;
    char g, h;

    // Operadores de asignación
    // C++ tiene los operadores de asignación: = (asignación simple), += (suma y asignación), -= (resta y asignación),
    // *= (multiplicación y asignación), /= (división y asignación) y %= (módulo y asignación).

    std::cout << "************************" << std::endl;
    std::cout << "OPERADORES DE ASIGNACIÓN" << std::endl;
    std::cout << "************************" << std::endl;

    std::cout << "" << std::endl;
    a = 12; // Asignación simple
    std::cout << "a = " << a << std::endl;
    a += 5; // Suma y asignación (a = a + 5)
    std::cout << "a += 5 -> a = " << a << std::endl;
    a -= 3; // Resta y asignación (a = a - 3)
    std::cout << "a -= 3 -> a = " << a << std::endl;
    a *= 2; // Multiplicación y asignación (a = a * 2)
    std::cout << "a *= 2 -> a = " << a << std::endl;
    a /= 4; // División y asignación (a = a / 4)
    std::cout << "a /= 4 -> a = " << a << std::endl;
    a %= 2; // Módulo y asignación (a = a % 2)
    std::cout << "a %= 2 -> a = " << a << std::endl;
    a++; // Incremento (a = a + 1)
    std::cout << "a++ -> a = " << a << std::endl;
    a--; // Decremento (a = a - 1)
    std::cout << "a-- -> a = " << a << std::endl;

    // Operaciones aritméticas
    // C++ tiene los operadores aritméticos básicos: + (suma), - (resta), * (multiplicación), / (división) y % (módulo).

    std::cout << "" << std::endl;
    std::cout << "**********************" << std::endl;
    std::cout << "OPERADORES ARITMÉTICOS" << std::endl;
    std::cout << "**********************" << std::endl;

    std::cout << "" << std::endl;
    std::cout << "1) Operaciones numeros enteros" << std::endl;
    a = 10, b = 5;
    std::cout << "a = " << a << " y b = " << b << std::endl;
    std::cout << "a + b = " << a + b << std::endl; // Suma
    std::cout << "a - b = " << a - b << std::endl; // Resta
    std::cout << "a * b = " << a * b << std::endl; // Multiplicación
    std::cout << "a / b = " << a / b << std::endl; // División
    std::cout << "a % b = " << a % b << std::endl; // Módulo
    
    std::cout << "" << std::endl;
    std::cout << "2) Operaciones numeros decimales" << std::endl;
    c = 10.5, d = 5.2;
    std::cout << "c = " << c << " y d = " << d << std::endl;
    std::cout << "c + d = " << c + d << std::endl; // Suma
    std::cout << "c - d = " << c - d << std::endl; // Resta
    std::cout << "c * d = " << c * d << std::endl; // Multiplicación
    std::cout << "c / d = " << c / d << std::endl; // División

    // Las operaciones aritméticas en c++ tienen un orden de precedencia, que es el siguiente:
    // 1. Paréntesis ()
    // 2. Exponentes (no existe un operador de potencia en C++, pero se puede usar la función pow() de la biblioteca <cmath>)
    // 3. Multiplicación *, División / y Módulo %
    // 4. Suma + y Resta -

    // Operadores de comparación
    // C++ tiene los operadores de comparación: == (igual a), != (diferente de), > (mayor que), < (menor que), 
    // >= (mayor o igual que) y <= (menor o igual que).
    
    std::cout << "" << std::endl;
    std::cout << "*************************" << std::endl;
    std::cout << "OPERADORES DE COMPARACIÓN" << std::endl;
    std::cout << "*************************" << std::endl;
    std::cout << "" << std::endl;

    e = (a > b); // ¿a es mayor que b?
    std::cout << "a es mayor que b: " << e << std::endl;
    e = (a < b); // ¿a es menor que b?
    std::cout << "a es menor que b: " << e << std::endl;
    e = (a == b); // ¿a es igual a b?
    std::cout << "a es igual a b: " << e << std::endl;
    e = (a != b); // ¿a es diferente de b?
    std::cout << "a es diferente de b: " << e << std::endl;
    e = (a >= b); // ¿a es mayor o igual que b?
    std::cout << "a es mayor o igual que b: " << e << std::endl;
    e = (a <= b); // ¿a es menor o igual que b?
    std::cout << "a es menor o igual que b: " << e << std::endl;

    // Operadores lógicos
    // C++ tiene los operadores lógicos: && (AND), || (OR) y ! (NOT).

    std::cout << "" << std::endl;
    std::cout << "******************" << std::endl;
    std::cout << "OPERADORES LÓGICOS" << std::endl;
    std::cout << "******************" << std::endl;
    std::cout << "" << std::endl;

    f = (a > b) && (c > d); // ¿a es mayor que b Y c es mayor que d? (operador AND)
    std::cout << "a es mayor que b Y c es mayor que d: " << f << std::endl;
    f = (a > b) || (c < d); // ¿a es mayor que b O c es menor que d? (operador OR)
    std::cout << "a es mayor que b O c es menor que d: " << f << std::endl;
    f = !(a > b); // NO a es mayor que b (operador NOT)
    std::cout << "NO a es mayor que b: " << f << std::endl;

    // Operadores de Bits
    /* C++ tiene los siguientes operadores de bits:
     * & (AND)
     * | (OR)
     * ^ (XOR)
     * ~ (NOT)
     * << (desplazamiento a la izquierda)
     * >> (desplazamiento a la derecha)
     */

    std::cout << "" << std::endl;
    std::cout << "******************" << std::endl;
    std::cout << "OPERADORES DE BITS" << std::endl;
    std::cout << "******************" << std::endl;
    std::cout << "" << std::endl;

    i = a & b; // AND bit a bit
    std::cout << "a AND b = " << i << std::endl;
    i = a | b; // OR bit a bit
    std::cout << "a OR b = " << i << std::endl;
    i = a ^ b; // XOR bit a bit
    std::cout << "a XOR b = " << i << std::endl;
    i = ~a; // NOT bit a bit
    std::cout << "NOT a = " << i << std::endl;
    i = a << 1; // Desplazamiento a la izquierda de a por 1 bit
    std::cout << "a desplazado a la izquierda por 1 bit = " << i << std::endl;
    i = a >> 1; // Desplazamiento a la derecha de a por 1 bit
    std::cout << "a desplazado a la derecha por 1 bit = " << i << std::endl;

    // En C++ NO existen operadores de pertenencia ni operadores de identidad como en Python, pero se pueden simular 
    // con operadores de comparación y lógicos.

    // Estructuras de control
    // C++ tiene las siguientes estructuras de control:
    // 1. Condicionales: if, if else, switch
    // 2. Bucles: for, while, do-while
    
    std::cout << "" << std::endl;
    std::cout << "**********************" << std::endl;
    std::cout << "ESTRUCTURAS DE CONTROL" << std::endl;
    std::cout << "**********************" << std::endl;
    std::cout << "" << std::endl;
    // Condicionales

    std::cout << "1) Condicionales" << std::endl;
    std::cout << "" << std::endl;
    std::cout << "1.1) if simple" << std::endl;
    std::cout << "" << std::endl;
    if (a > b) 
        std::cout << "a es mayor que b" << std::endl; // if simple sin llaves, solo para una línea de código
    
    std::cout << "" << std::endl;
    std::cout << "1.2) if simple con llaves" << std::endl;
    std::cout << "" << std::endl;
    if (a > b) {
        std::cout << "a es mayor que b" << std::endl; // if simple con llaves, para una o más líneas de código
    }

    std::cout << "" << std::endl;
    std::cout << "1.3) if else" << std::endl;
    std::cout << "" << std::endl;
    if (a > b) {
        std::cout << "a es mayor que b" << std::endl; // if else con llaves, para una o más líneas de código
    } else {
        std::cout << "a no es mayor que b" << std::endl;
    }

    std::cout << "" << std::endl;
    std::cout << "1.4) Switch" << std::endl;
    std::cout << "Ingrese un valor entero para a: ";
    std::cin >> a; // Leer un valor entero para a desde la entrada estándar

    switch (a) { // switch para múltiples condiciones basadas en el valor de a
        case 0:
            std::cout << "a es cero" << std::endl;
            break;
        case 1:
            std::cout << "a es uno" << std::endl;
            break;
        case 10:
            std::cout << "a es diez" << std::endl;
            break;
        default:
            std::cout << "a no es ni cero, ni uno, ni diez" << std::endl;
    }
    
    // Bucles
    std::cout << "" << std::endl;
    std::cout << "2) Bucles" << std::endl;
    std::cout << "" << std::endl;
    a=10; // Asignar a el valor 10 para los bucles
    std::cout << "2.1) Bucle while" << std::endl;
    while (a > 0) { // Bucle while, se ejecuta mientras a sea mayor que 0
        std::cout << "a = " << a << std::endl;
        a--; // Decrementar a en 1
    }

    std::cout << "" << std::endl;
    std::cout << "2.2) Bucle do-while" << std::endl;
    do { // Bucle do-while, se ejecuta al menos una vez y luego se repite mientras a sea mayor que 0
        std::cout << "a = " << a << std::endl;
        a++; // Incrementar a en 1
    } while (a < 5);

    std::cout << "" << std::endl;
    std::cout << "2.3) Bucle for" << std::endl;
    for (i = 0; i < 5; i++) { // Bucle for, se ejecuta desde i = 0 hasta i < 5, incrementando i en 1 en cada iteración
        std::cout << "i = " << i << std::endl;
    }

    std::cout << "" << std::endl;
    std::cout << "****************" << std::endl;
    std::cout << "DIFICULTAD EXTRA" << std::endl;
    std::cout << "****************" << std::endl;
    std::cout << "" << std::endl;

    for (i = 10; i <= 55; i++) {
        if (i % 2 == 0 && i != 16 && i % 3 != 0 || i == 55) { 
            // Imprimir los números pares entre 10 y 55, excepto el 16 y los múltiplos de 3
            std::cout << i << std::endl;
        }    
    }
}